Last Updated: 05/12/2025 by Ashley Gin & Elif Kulaksizoglu (20.440 Term Project)

# **Characterizing Natural Killer Cell Subsets Contributing to Antiviral Memory**

## **Overview**
This project aims to investigate antiviral natural killer cell memory by characterizing differential gene expression of naive and adaptive NK cell subsets following infection by human cytomegalovirus (HCMV). 

This repository contains human patient scRNA-seq data, the pipeline script required to clean the data and generate a labeled UMAP, and example PDF figure outputs. This repository can be used to generate all figures in the final draft of the NK cell report. The pipeline performs preprocessing, normalization, dimensionality reduction (PCA), UMAP clustering, differential gene expression, and gene set enrichment analysis (GSEA).

The code was adapted from: RÃ¼ckert T, Lareau CA, Mashreghi MF, Ludwig LS, Romagnani C. Clonal expansion and epigenetic inheritance of long-lasting NK cell memory. Nat Immunol. 2022 Nov;23(11):1551-1563. doi: 10.1038/s41590-022-01327-7. Epub 2022 Oct 26. PMID: 36289449; PMCID: PMC9663309.

## **Data**
The NK cell scRNA-seq data was collected from primary NK cells isolated from donor peripheral blood for patients infected and uninfected with HCMV (denoted HCMVpos and HCMVneg, respectively in the data folders). Patient HCMV status was determined by serology. 

Data contains 3 HCMV+ patients and 2 HCMV- patients for healthy control.
Each sample contains the barcodes, features, and matrix files required to generate the Seurat object for downstream UMAP clustering and analysis.

Adapted from: https://github.com/timorueckert/Clonal_NK 

## **Folder Structure**
* Code: contains the R script pipeline required to clean data and analyze HCMV+ and HCMV- scRNA-seq data
* Data: contains 3 HCMV+ and 2 HCMV- samples, with each sample containing individual barcode, feature, and matrix files
* Outputs: contains example PDF outputs generated by the code

## **Installation**
* This code can be run via R Version 4.4.3 in RStudio. All required packages are initialized in the provided script (packages: dplyr, parallel, Seurat, scater, scran, purrr, viridis, ggplot2, RColorBrewer, Matrix, data.table, stringr, future, msigdbr, fgsea).
